### PR TITLE
Cache whether RUST_BACKTRACE is enabled in a relaxed atomic static.

### LINKED ISSUE #207: env::var_os("RUST_BACKTRACE") should only be called once.
The caching can be done with just an `AtomicBool` (thanks, @edef1c!).
I'm seeing something like a hundred thousand allocations from this, per second, and C calls, of course, and none of those are really needed (they might be small but they're profiling noise).

### PR DESCRIPTION
**Pull request by eddyb**
Fixes #207. Inspired by [libfringe](https://github.com/edef1c/libfringe/blob/87452b58e0a4cbef4973e81419d719b0b8fc298b/src/stack/os/sys.rs#L63-L71) (thanks, @edef1c!).

### CODE REVIEW

**Comment by eddyb** [comment]
Apparently [there's a test that changes `RUST_BACKTRACE` at runtime](https://github.com/rust-lang-nursery/error-chain/blob/527bea669a8b988b6da2f204525614b7e4e5f3ca/tests/tests.rs#L227).
The intention seems to be to check a few `RUST_BACKTRACE` values, shouldn't it be done by running a different executable via `std::process`?

**Comment by Yamakaky** [comment]
It's an edge case, I think a subprocess would be fine.

**Comment by Arnavion** [comment]
Perhaps split it up into separate tests (one for each `RUST_BACKTRACE` value rather than actual subprocesses?

**Comment by faern** [comment]
Would it not make sense to implement this with `std::sync::Once` instead of manually having to encode whether or not the environment variable has been loaded or not? But maybe that has noticeably worse performance?

**Comment by faern** [comment]
I have created a PR on @eddyb 's fork that fixes the failing test by running it as a subprocess. https://github.com/eddyb/error-chain/pull/1

I have verified it works on Linux, but I have not tested other platforms. I wrote the code with a special case for Windows, but I have not verified that it actually runs.

**Comment by eddyb** [comment]
Thanks, @faern! ping @Yamakaky

File: src/bin/has_backtrace.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+//! Exits with exit code 0 if backtraces are disabled and 1 if they are enabled.
+//! Used by tests to make sure backtraces are available when they should be.
```
Comment (Yamakaky) [inline src/bin/has_backtrace.rs]: Maybe add that this bin should not be used outside of the tests?

File: tests/tests.rs [hunk 2]
```diff
@@ -200,35 +200,29 @@ fn empty() {
...
-        errors {
-            MyError
-        }
-    }
-
-    let original_value = env::var_os("RUST_BACKTRACE");
+    let cmd_path = if cfg!(windows) {
+        Path::new("./target/debug/has_backtrace.exe")
+    } else {
+        Path::new("./target/debug/has_backtrace")
```
Comment (eddyb) [inline tests/tests.rs]: I would use `cargo run --bin` and maybe put it somewhere `cargo` wouldn't pick up by default. I can try that if it sounds good.

**Comment by Yamakaky** [comment]
Nice! Could you update the changelog?

**Comment by faern** [comment]
I added PR for updating change log: https://github.com/eddyb/error-chain/pull/2

**Comment by eddyb** [comment]
I still have to play with `cargo run`, unless @faern gets to it first 😆

**Comment by faern** [comment]
What is it that you are trying to solve? Moving `has_backtrace.rs` out of `bin/` so it is not automatically picked up and built by `cargo test`? `example/` does not sound like an optimal place to put it IMO. `example/` is also automatically picked up by `cargo test`, plus users looking there expect to see examples of how to use the library.

**Comment by faern** [comment]
Manually compiling it has the downside of us not knowing how the user has configured their toolchain. Just running `rustc ...` might not be on the PATH or might not be what the user want to run. Maybe they have invoked the tests with `cargo +beta test` which is something a hardcoded `cargo ...`/`rustc ...` would miss.

**Comment by eddyb** [comment]
Hmm in those cases I don't know how to actually have a separate process. Then again, this is a test so if `cargo run` doesn't work...

**Comment by faern** [comment]
I still didn't get what the problem we are trying to solve is? Is it too ugly to have it in `bin/`? Even if we rename it to `test_has_backtrace.rs` or `internal_test_helper__do_not_use__has_backtrace.rs`? :D

**Comment by eddyb** [comment]
It *doesn't* work. Look at the test failure. It worked for you only because you built it manually.

**Comment by faern** [comment]
Hmm. What? `cargo clean && cargo test` works for me. I didn't build anything manually. There is only one build failing, and it looks like that is because of some syntax error in the backtrace crate.

**Comment by eddyb** [comment]
@faern Huh, my bad, guess it... fixed itself? Really weird since the last failure I saw was not finding that executable. I'll try to rebase then.

**Comment by eddyb** [comment]
I... rebased, why does GitHub still complain about conflicts?

**Comment by faern** [comment]
It looks like we have to bump the minimum required version from `1.13.0` to `1.14.0`. Looks like `backtrace 0.3.3` won't build on Rust `1.13.0`.

**Comment by faern** [comment]
Ping @Yamakaky :)

**Comment by Yamakaky** [comment]
Yeah, it's good !

Good coop btw, keeping an eye on it ;)

**Comment by Yamakaky** [comment]
Are you good for merge?

**Comment by eddyb** [comment]
I think so.

File: src/lib.rs [hunk 3]
```diff
@@ -593,9 +593,29 @@ impl<'a> Iterator for Iter<'a> {
...
+            let enabled = match std::env::var_os("RUST_BACKTRACE") {
+                Some(ref val) if val != "0" => true,
+                _ => false
+            };
+            let encoded = ((enabled as usize) << 1) | 1;
+            BACKTRACE_ENABLED_CACHE.store(encoded, Ordering::Relaxed);
+            enabled
+        }
+        encoded => (encoded >> 1) != 0
+    };
```
Comment (Yamakaky) [inline src/lib.rs]: I'm just thinking this piece of code could be reused in others libs. Something like `LazyBool` or something like that.

**Comment by Yamakaky** [comment]
Manually merged. Thank you two!

File: src/lib.rs (hunk 3 above)
Comment (faern) [reply src/lib.rs]: I agree that this could me generalized. Or at least refactored into a separate function, since it's fairly large.
